File: IsraelHiking.Web/android/app/src/main/java/com/mapeak/car/CarMapScreen.java
package com.mapeak.car;

import androidx.annotation.NonNull;
import androidx.car.app.CarContext;
import androidx.car.app.Screen;
import androidx.car.app.model.Action;
import androidx.car.app.model.ActionStrip;
import androidx.car.app.model.CarIcon;
import androidx.car.app.model.DateTimeWithZone;
import androidx.car.app.model.Distance;
import androidx.car.app.model.Template;
import androidx.car.app.navigation.model.NavigationTemplate;
import androidx.car.app.navigation.model.TravelEstimate;
import androidx.core.graphics.drawable.IconCompat;
import androidx.lifecycle.DefaultLifecycleObserver;
import androidx.lifecycle.LifecycleOwner;

import com.mapeak.R;

import java.util.Locale;
import java.util.TimeZone;

public class CarMapScreen extends Screen implements CarMessageBus.CarEventListener, DefaultLifecycleObserver {

    private final CarContext carContext;
    private final CarMapRenderer carMapRenderer;

    private Double remainingMeters = null;
    private Long remainingSeconds = null;
    private String units = "metric";

    public CarMapScreen(CarContext carContext, CarMapRenderer carMapRenderer) {
        super(carContext);
        this.carContext = carContext;
        this.carMapRenderer = carMapRenderer;
        getLifecycle().addObserver(this);
    }

    @Override
    public void onCreate(@NonNull LifecycleOwner owner) {
        CarMessageBus.getInstance().registerListener(this);
    }

    @Override
    public void onDestroy(@NonNull LifecycleOwner owner) {
        CarMessageBus.getInstance().unregisterListener(this);
    }

    @Override
    public void onCarEvent(CarMessageBus.CarEvent event) {
        if (!CarMessageBus.EVENT_STATISTICS.equals(event.actionId())) {
            return;
        }
        var payload = event.payload();
        if (payload == null) {
            remainingMeters = null;
            remainingSeconds = null;
        } else {
            remainingMeters = payload.optDouble("remainingMeters");
            remainingSeconds = payload.optLong("remainingSeconds");
            units = payload.optString("units", "metric");
        }
        invalidate();
    }

    @NonNull
    @Override
    public Template onGetTemplate() {
        NavigationTemplate.Builder templateBuilder = new NavigationTemplate.Builder();
        templateBuilder.setActionStrip(buildActionStrip().build());
        if (carContext.getCarAppApiLevel() >= 2) {
            templateBuilder.setMapActionStrip(buildMapActionStrip(carMapRenderer).build());
        }
        TravelEstimate estimate = buildTravelEstimate();
        if (estimate != null) {
            templateBuilder.setDestinationTravelEstimate(estimate);
        }
        return templateBuilder.build();
    }

    private TravelEstimate buildTravelEstimate() {
        if (remainingMeters == null || remainingSeconds == null) {
            return null;
        }
        boolean isImperial = "imperial".equals(units);
        Distance remainingDistance = isImperial
                ? Distance.create(remainingMeters / 1609.344, Distance.UNIT_MILES)
                : Distance.create(remainingMeters / 1000.0, Distance.UNIT_KILOMETERS);
        DateTimeWithZone arrivalTime = DateTimeWithZone.create(System.currentTimeMillis() + remainingSeconds * 1000,
                TimeZone.getDefault());
        return new TravelEstimate.Builder(remainingDistance, arrivalTime)
                .setRemainingTimeSeconds(remainingSeconds)
                .build();
    }

    private ActionStrip.Builder buildActionStrip() {
        ActionStrip.Builder actionStripBuilder = new ActionStrip.Builder();

        actionStripBuilder.addAction(
                new Action.Builder()
                        .setIcon(
                                new CarIcon.Builder(
                                        IconCompat.createWithResource(
                                                carContext,
                                                R.drawable.ic_menu))
                                        .build())
                        .build());

        return actionStripBuilder;
    }

    private ActionStrip.Builder buildMapActionStrip(CarMapRenderer carMapRenderer) {
        ActionStrip.Builder actionStripBuilder = new ActionStrip.Builder();

        actionStripBuilder.addAction(Action.PAN);

        actionStripBuilder.addAction(
                new Action.Builder()
                        .setIcon(
                                new CarIcon.Builder(
                                        IconCompat.createWithResource(
                                                carContext,
                                                R.drawable.ic_zoom_in))
                                        .build())
                        .setOnClickListener(carMapRenderer::zoomInFromButton)
                        .build());

        actionStripBuilder.addAction(
                new Action.Builder()
                        .setIcon(
                                new CarIcon.Builder(
                                        IconCompat.createWithResource(
                                                carContext,
                                                R.drawable.ic_zoom_out))
                                        .build())
                        .setOnClickListener(carMapRenderer::zoomOutFromButton)
                        .build());

        return actionStripBuilder;
    }
}

File: IsraelHiking.Web/android/app/src/main/java/com/mapeak/car/CarMapScreen.kt
package com.mapeak.car

import android.location.Location
import androidx.annotation.DrawableRes
import androidx.car.app.CarContext
import androidx.car.app.Screen
import androidx.car.app.model.Action
import androidx.car.app.model.ActionStrip
import androidx.car.app.model.CarColor
import androidx.car.app.model.CarIcon
import androidx.car.app.model.DateTimeWithZone
import androidx.car.app.model.Distance
import androidx.car.app.model.Template
import androidx.car.app.navigation.model.NavigationTemplate
import androidx.car.app.navigation.model.TravelEstimate
import androidx.core.graphics.drawable.IconCompat
import androidx.lifecycle.DefaultLifecycleObserver
import androidx.lifecycle.LifecycleOwner
import com.mapeak.R
import java.util.TimeZone
import org.json.JSONException

class CarMapScreen(
        private val carContext: CarContext,
        private val carMapRenderer: CarMapRenderer,
        private val navigation: CarNavigation,
        private val pendingNavigationQuery: String? = null
) : Screen(carContext), CapacitorStore.Listener, DefaultLifecycleObserver {

    private val store: CapacitorStore = CapacitorStore.get(carContext)
    private var routes: List<CarRouteData> = emptyList()
    private var statistics: CarStatistics? = null
    private var units: String = DEFAULT_UNITS

    init {
        lifecycle.addObserver(this)
    }

    /**
     * On a "navigate to X" intent ([pendingNavigationQuery] set), search is opened on top of the map
     * so the driver sees results/route for X while the map remains the root they return to.
     */
    override fun onCreate(owner: LifecycleOwner) {
        store.addListener(this)
        navigation.onNavigationInfoChanged = { invalidate() }
        loadRoutes()
        loadUnits()
        if (!pendingNavigationQuery.isNullOrBlank()) {
            screenManager.push(CarSearchScreen(carContext, pendingNavigationQuery))
        }
    }

    override fun onDestroy(owner: LifecycleOwner) {
        store.removeListener(this)
        navigation.onNavigationInfoChanged = null
    }

    override fun onCarStoreUpdated(key: String) {
        when (key) {
            CarStoreKeys.LOCATION -> recomputeStatistics()
            CarStoreKeys.ROUTE -> {
                loadRoutes()
                recomputeStatistics()
            }
            CarStoreKeys.CONFIG -> {
                loadUnits()
                invalidate()
            }
        }
    }

    private fun recomputeStatistics() {
        val location: Location? = store.getTransient(CarStoreKeys.LOCATION)
        val next = if (location == null) null else CarRouteCalculator.computeStatistics(routes, location)
        if (next != statistics) {
            statistics = next
            invalidate()
        }
    }

    private fun loadRoutes() {
        routes =
                try {
                    CarRouteData.listFromJson(store.load(CarStoreKeys.ROUTE))
                } catch (_: JSONException) {
                    emptyList()
                }
    }

    private fun loadUnits() {
        units = store.load(CarStoreKeys.CONFIG)?.optString("units") ?: DEFAULT_UNITS
    }

    override fun onGetTemplate(): Template {
        val templateBuilder = NavigationTemplate.Builder().setActionStrip(buildActionStrip())
        if (carContext.carAppApiLevel >= 2) {
            templateBuilder.setMapActionStrip(buildMapActionStrip())
        }
        buildTravelEstimate()?.let { templateBuilder.setDestinationTravelEstimate(it) }
        navigation.navigationInfo?.let { templateBuilder.setNavigationInfo(it) }
        return templateBuilder.build()
    }

    private fun buildTravelEstimate(): TravelEstimate? {
        val stats = statistics ?: return null
        val remainingDistance =
                if (units == UNIT_IMPERIAL) {
                    Distance.create(stats.remainingMeters / METERS_PER_MILE, Distance.UNIT_MILES)
                } else {
                    Distance.create(
                            stats.remainingMeters / METERS_PER_KILOMETER,
                            Distance.UNIT_KILOMETERS
                    )
                }
        val arrivalTime =
                DateTimeWithZone.create(
                        System.currentTimeMillis() + stats.remainingSeconds * 1000,
                        TimeZone.getDefault()
                )
        return TravelEstimate.Builder(remainingDistance, arrivalTime)
                .setRemainingTimeSeconds(stats.remainingSeconds)
                .build()
    }

    private fun buildActionStrip(): ActionStrip =
            ActionStrip.Builder()
                    .addAction(
                            iconAction(R.drawable.ic_search, false) {
                                screenManager.push(CarSearchScreen(carContext, null))
                            }
                    )
                    .build()

    private fun buildMapActionStrip(): ActionStrip =
            ActionStrip.Builder()
                    .addAction(Action.PAN)
                    .addAction(
                            iconAction(R.drawable.ic_zoom_in) { carMapRenderer.zoomInFromButton() }
                    )
                    .addAction(
                            iconAction(R.drawable.ic_zoom_out) {
                                carMapRenderer.zoomOutFromButton()
                            }
                    )
                    .addAction(
                            iconAction(R.drawable.ic_recenter) {
                                carMapRenderer.recenterFromButton()
                            }
                    )
                    .build()

    /**
     * Builds an action button. The icon uses the default tint so it adapts to the host's day/night
     * theme and the glyph stays legible in both modes.
     */
    private fun iconAction(
            @DrawableRes iconRes: Int,
            persist: Boolean = true,
            onClick: (() -> Unit)? = null,
    ): Action {
        val builder =
                Action.Builder()
                        .setIcon(
                                CarIcon.Builder(IconCompat.createWithResource(carContext, iconRes))
                                        .setTint(CarColor.DEFAULT)
                                        .build()
                        )
        if (persist && carContext.carAppApiLevel >= PERSISTENT_ACTION_MIN_API) {
            builder.setFlags(Action.FLAG_IS_PERSISTENT)
        }
        if (onClick != null) {
            builder.setOnClickListener { onClick() }
        }
        return builder.build()
    }

    companion object {
        private const val DEFAULT_UNITS = "metric"
        private const val UNIT_IMPERIAL = "imperial"
        private const val METERS_PER_KILOMETER = 1000.0
        private const val METERS_PER_MILE = 1609.344
        private const val PERSISTENT_ACTION_MIN_API = 5
    }
}
